File: LeetCode/801-900/Problem885_SpiralMatrixIII/SpiralMatrixIII.py
from typing import List

class Solution:
    def spiralMatrixIII(self, rows: int, cols: int, rStart: int, cStart: int) -> List[List[int]]:
        total = rows * cols
        result = [[rStart, cStart]]
        
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        r, c = rStart, cStart
        step = 1
        count = 1

        def in_bounds(r, c):
            return 0 <= r < rows and 0 <= c < cols

        while count < total:
            for i in range(4):
                dx, dy = directions[i]
                for _ in range(step):
                    r += dx
                    c += dy
                    if in_bounds(r, c):
                        result.append([r, c])
                        count += 1
                        if count == total:
                            return result
                if i % 2 == 1:
                    step += 1

        return result

File: LeetCode/801-900/Problem885_SpiralMatrixIII/SpiralMatrixIII.cpp
class Solution {
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int rStart, int cStart) {
        const int total = rows * cols;
        vector<vector<int>> result;
        result.reserve(total);

        int dirs[4][2] = {{0,1}, {1,0}, {0,-1}, {-1,0}};

        auto inBounds = [&](int r, int c) {
            return static_cast<unsigned>(r) < static_cast<unsigned>(rows) &&
                   static_cast<unsigned>(c) < static_cast<unsigned>(cols);
        };

        int r = rStart, c = cStart, step = 1, count = 1;
        result.push_back({r, c});

        while (count < total) {
            for (int i = 0; i < 4; ++i) {
                int dx = dirs[i][0], dy = dirs[i][1];
                for (int j = 0; j < step; ++j) {
                    r += dx;
                    c += dy;
                    if (inBounds(r, c)) {
                        result.push_back({r, c});
                        if (++count == total) return result;
                    }
                }
                if (i % 2 == 1) ++step;
            }
        }

        return result;
    }
};
